### DISCUSSION TITLE
Combat info logging

### DISCUSSION THREAD

**Discussion by pruuk** [evennia/evennia#2408] [Evennia Questions] [answered]
I've got my combat system up and working. It has a lot of factors that I can easily tweak to balance combat between different styles (unarmed strikes, unarmed grappling, melee weapons of different types, ranged combat, etc). Up to now, I've been testing and using the msg function to show these factors, but it gets really spammy and isn't as useful as a real combat log.

What do you think is the best way to go about building a combat logger? Using the evennia log mechanisms? Storing it on an object or a script? Writing it to a file somewhere else? I'd like to be able to access the info outside of the game and manipulate it with numpy and visualize it with matplotlib so I can make balance tweaks.

Thanks for any advice/ideas you might have!

**Comment by Griatch** [comment, marked as answer]
It depends on what you want your combat logger to do. It sounds like you primarily want it for development/debugging. For this I'd make a custom combat file logger using `evennia.utils.logger.log_file`. It's simple enough to use:

```
from evennia.utils.logger import log_file

log_file("Pruuk started an attack!", filename="combat.log")
```

The file `combat.log` will appear in `mygame/server/logs/` and will be automatically tagged with the current time. This is easy enough to use for your own development tests. 

Of course, if your goal is to allow users to actually _read_ from the log, you need to take a page from the way Channel `history` command works. You will find that `evennia.utils.logger.tail_log_file` is very useful for easily providing a section of the log file for people to read (of course, you'd likely need separate files for each combat for this to make sense, and then using some database-based solution like storing things on a Script (like the one created in the Turn-based combat tutorial) may be more suitable.

**Reply by pruuk** [reply]
It looks like the logger solution should work just fine for what I need. I'll give that a shot and see how it works for me.

Thanks for replying back so quickly.

**Reply by pruuk** [reply]
That worked beautifully. Thanks again!

**Comment by TehFamine** [comment]
Just some food for thought.

Evennia is Python and there are great Python options for Azure and AWS if you happen to host there. I make extensive logs for my combat to do analytics with later. One scenario is to do similar with logging like Griatch said, but instead pushing the log to DynamoDB or something similar to that with the boto3 package. That or pushing up to CloudWatch or if in Azure, Application Insights. Then all of your logs are processed externally and can be queried for reporting. For example, pushed into Power BI, Google Sheets, Excel, whatever.

**Reply by pruuk** [reply]
I haven't looked into my hosting options yet other than reading through the Evennia help files. I already got frustrated with a design choice I made on the third version of my MUD and started re-writing a 4th time. :)

It'll be a whole new set of skills to learn once I get to the point where I'm ready to host the MUD. Thank you for the tips; I'll look into the options you've mentioned.
